### ISSUE TITLE
DISABLED test_backward_node_failure (__main__.TensorPipeDistAutogradTestWithSpawn)

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#39855]
https://app.circleci.com/pipelines/github/pytorch/pytorch/179019/workflows/89ef9bec-c13c-4d81-a817-5de5fed67cdf/jobs/5772385/steps https://app.circleci.com/pipelines/github/pytorch/pytorch/179483/workflows/b1faffb1-687b-4000-bae1-37fa2c93d7f5/jobs/5787474/steps

```
Jun 10 18:19:54   test_backward_node_failure (__main__.TensorPipeAgentDistAutogradTestWithSpawn) ... [W tensorpipe_agent.cpp:315] RPC agent for worker3 encountered error when reading incoming request: pipe closed
Jun 10 18:19:54 [W tensorpipe_agent.cpp:315] RPC agent for worker1 encountered error when reading incoming request: pipe closed
Jun 10 18:19:54 [E container.cpp:248] Could not release Dist Autograd Context on node 0: pipe closed
Jun 10 18:19:54 [E container.cpp:248] Could not release Dist Autograd Context on node 2: pipe closed
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker0 encountered error when writing outgoing request: EPIPE: broken pipe
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker2 encountered error when writing outgoing request: ECONNREFUSED: connection refused
Jun 10 18:19:54 [W tensorpipe_agent.cpp:315] RPC agent for worker0 encountered error when reading incoming request: EOF: end of file
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker0 encountered error when writing outgoing request: ECONNREFUSED: connection refused
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker0 encountered error when writing outgoing request: ECONNREFUSED: connection refused
Jun 10 18:19:54 [W tensorpipe_agent.cpp:283] RPC agent for worker0 encountered error when writing outgoing response: EOF: end of file
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker0 encountered error when writing outgoing request: EPIPE: broken pipe
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker2 encountered error when writing outgoing request: ECONNREFUSED: connection refused
Jun 10 18:19:54 [W tensorpipe_agent.cpp:315] RPC agent for worker2 encountered error when reading incoming request: EPIPE: broken pipe
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker0 encountered error when writing outgoing request: ECONNREFUSED: connection refused
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker0 encountered error when writing outgoing request: EPIPE: broken pipe
Jun 10 18:19:54 [E container.cpp:248] Could not release Dist Autograd Context on node 3: ECONNREFUSED: connection refused
Jun 10 18:19:54 [E container.cpp:248] Could not release Dist Autograd Context on node 1: EPIPE: broken pipe
Jun 10 18:19:54 [W tensorpipe_agent.cpp:283] RPC agent for worker2 encountered error when writing outgoing response: EPIPE: broken pipe
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker2 encountered error when writing outgoing request: EPIPE: broken pipe
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker2 encountered error when writing outgoing request: EPIPE: broken pipe
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker2 encountered error when writing outgoing request: ECONNREFUSED: connection refused
Jun 10 18:19:54 [W tensorpipe_agent.cpp:439] RPC agent for worker2 encountered error when writing outgoing request: EPIPE: broken pipe
Jun 10 18:21:33 Timing out after 100 seconds and killing subprocesses.
Jun 10 18:21:33 ERROR (100.075s)
```

```
Jun 10 18:24:42 ======================================================================
Jun 10 18:24:42 ERROR [100.075s]: test_backward_node_failure (__main__.TensorPipeAgentDistAutogradTestWithSpawn)
Jun 10 18:24:42 ----------------------------------------------------------------------
Jun 10 18:24:42 Traceback (most recent call last):
Jun 10 18:24:42   File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 204, in wrapper
Jun 10 18:24:42     self._join_processes(fn)
Jun 10 18:24:42   File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 306, in _join_processes
Jun 10 18:24:42     self._check_return_codes(elapsed_time)
Jun 10 18:24:42   File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 344, in _check_return_codes
Jun 10 18:24:42     raise RuntimeError('Process {} terminated or timed out after {} seconds'.format(i, elapsed_time))
Jun 10 18:24:42 RuntimeError: Process 2 terminated or timed out after 100.05913996696472 seconds
Jun 10 18:24:42 
Jun 10 18:24:42 ----------------------------------------------------------------------
```

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @gqchen @pearu @nikitaved @soulitzer @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @aazzolini @rohan-varma @jjlilley @osalpekar @jiayisuse @mrzzd @lw @beauby @agolynski @SciPioneer @H-Huang @cbalioglu @SsnL @xush6528

**Comment by pritamdamania87** [comment]
Haven't been able to repro this locally. Closing this issue out to see if this is resolved in master already.

**Comment by rohan-varma** [comment]
It looks like this test failed on one of my recent CI jobs: https://app.circleci.com/pipelines/github/pytorch/pytorch/181976/workflows/fa8529f3-017a-481f-a864-1a95f2f24712/jobs/5879192/steps, and I verified the fix is included in my diff.

**Comment by lw** [comment]
Reopening because the issue persists.

I've made the logging in RPC even more aggressive (#40152) and then logged into a CircleCI machine to try to debug this issue. I ran with all RPC logging enabled, at its most verbose setting, and got this trace: https://gist.github.com/lw/0f2ed26218c0de2c68b1a0630d288efc. At a first analysis, that trace seems to suggest that everything went according to plan: all operations terminated, all work was shut down.

So then I attached GDB to a live stuck process, and got this:
```
(gdb) info proc
process 27334
warning: target file /proc/27334/cmdline contained unexpected null characters
cmdline = '/opt/conda/bin/python'
cwd = '/var/lib/jenkins/workspace/test'
exe = '/opt/conda/bin/python3.8'
```
```
$ cat /proc/27334/cmdline
/opt/conda/bin/python -c from multiprocessing.spawn import spawn_main; spawn_main(tracker_fd=5, pipe_handle=11) --multiprocessing-fork
```
```
(gdb) info threads
  Id   Target Id         Frame
* 1    LWP 27334 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  2    LWP 27340 "python" 0x00007fbb85616bb7 in epoll_wait () from target:/lib/x86_64-linux-gnu/libc.so.6
  3    LWP 27362 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  4    LWP 27363 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  5    LWP 27364 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  6    LWP 27365 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  7    LWP 27366 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  8    LWP 27368 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  9    LWP 27371 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  10   LWP 27374 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  11   LWP 27380 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  12   LWP 27386 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  13   LWP 27391 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  14   LWP 27399 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  15   LWP 27403 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  16   LWP 27409 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  17   LWP 27412 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  18   LWP 27413 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  19   LWP 27414 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
  20   LWP 27415 "python" 0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
```
```

(gdb) thread apply all bt
Thread 1 (LWP 27334):
#0  0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007fbb72f5e4cb in __gthread_cond_wait (__mutex=<optimized out>, __cond=<optimized out>)
    at /home/nwani/m3/conda-bld/compilers_linux-64_1560109574129/work/.build/x86_64-conda_cos6-linux-gnu/build/build-cc-gcc-final/x86_64-conda_cos6-linux-gnu/libstdc++-v3/include/x86_64-conda_cos6-linux-gnu/bits/gthr-default.h:878
#2  std::condition_variable::wait (this=<optimized out>, __lock=...)
    at /home/nwani/m3/conda-bld/compilers_linux-64_1560109574129/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/src/c++11/condition_variable.cc:53
#3  0x00007fbb7698d153 in torch::autograd::Engine::~Engine() () from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#4  0x00007fbb85538041 in ?? () from target:/lib/x86_64-linux-gnu/libc.so.6
#5  0x00007fbb8553813a in exit () from target:/lib/x86_64-linux-gnu/libc.so.6
#6  0x000055ff247c9239 in Py_Exit (sts=0) at /tmp/build/80754af9/python_1578519979997/work/Python/pylifecycle.c:2299
#7  0x000055ff247c926d in handle_system_exit () at /tmp/build/80754af9/python_1578519979997/work/Python/pythonrun.c:658
#8  0x000055ff247c92c4 in _PyErr_PrintEx () at /tmp/build/80754af9/python_1578519979997/work/Python/pythonrun.c:668
#9  0x000055ff247cc176 in PyRun_SimpleStringFlags () at /tmp/build/80754af9/python_1578519979997/work/Python/pythonrun.c:462
#10 0x000055ff247cc2e5 in pymain_run_command (cf=0x7ffc725ebcf8, command=<optimized out>) at /tmp/build/80754af9/python_1578519979997/work/Modules/main.c:264
#11 pymain_run_python (exitcode=0x7ffc725ebcf0) at /tmp/build/80754af9/python_1578519979997/work/Modules/main.c:562
#12 Py_RunMain () at /tmp/build/80754af9/python_1578519979997/work/Modules/main.c:650
#13 0x000055ff247cc6e9 in Py_BytesMain () at /tmp/build/80754af9/python_1578519979997/work/Modules/main.c:1082
#14 0x00007fbb85516b97 in __libc_start_main () from target:/lib/x86_64-linux-gnu/libc.so.6
#15 0x000055ff2476cc13 in _start () at ../sysdeps/x86_64/elf/start.S:103

Thread 2 (LWP 27340):
#0  0x00007fbb85616bb7 in epoll_wait () from target:/lib/x86_64-linux-gnu/libc.so.6
#1  0x00007fbb77b3dd72 in gloo::transport::tcp::Loop::run() () from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#2  0x00007fbb72f6219d in std::execute_native_thread_routine (__p=0x55ff269d31f0)
    at /home/nwani/m3/conda-bld/compilers_linux-64_1560109574129/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/src/c++11/thread.cc:80
#3  0x00007fbb858ed6db in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#4  0x00007fbb8561688f in clone () from target:/lib/x86_64-linux-gnu/libc.so.6

Thread 3 (LWP 27362):
#0  0x00007fbb858f39f3 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007fbb72f5e4cb in __gthread_cond_wait (__mutex=<optimized out>, __cond=<optimized out>)
    at /home/nwani/m3/conda-bld/compilers_linux-64_1560109574129/work/.build/x86_64-conda_cos6-linux-gnu/build/build-cc-gcc-final/x86_64-conda_cos6-linux-gnu/libstdc++-v3/include/x86_64-conda_cos6-linux-gnu/bits/gthr-default.h:878
#2  std::condition_variable::wait (this=<optimized out>, __lock=...)
    at /home/nwani/m3/conda-bld/compilers_linux-64_1560109574129/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/src/c++11/condition_variable.cc:53
#3  0x00007fbb7cfdeb81 in c10d::ProcessGroupGloo::runLoop(int) () from target:/opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_python.so
#4  0x00007fbb72f6219d in std::execute_native_thread_routine (__p=0x55ff26a0abc0)
    at /home/nwani/m3/conda-bld/compilers_linux-64_1560109574129/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/src/c++11/thread.cc:80
#5  0x00007fbb858ed6db in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#6  0x00007fbb8561688f in clone () from target:/lib/x86_64-linux-gnu/libc.so.6
```
(and all other threads looking like thread 3)

TL;DR: The process seems to get stuck during exit, while destroying static variables. The specific line holding up the main thread seems to be this one:
https://github.com/pytorch/pytorch/blob/44c7a2ab69293562f33cc0b368501bfbfbd937b7/torch/csrc/autograd/engine.cpp#L233
It is not clear where that Gloo process group is coming from but it may be the global default one, which perhaps gets destroyed after the autograd engine?

I tried to inspect the autograd engine to find out what it thought the value of `non_reentrant_device_thread_count_` was but the build was lacking debug symbols and GDB wouldn't let me easily do that. I could also instrument that code to add a few prints and re-run the CI, but perhaps someone familiar with autograd should take a look at this first?

**Comment by ezyang** [comment]
This comment near the line you quoted seems relevant:

```

    // Do not wait for termination of global threads on Windows
    // Because CRT terminates DLL threads before calling
    // global object destructors
#if !defined(_WIN32) || !defined(C10_BUILD_SHARED_LIBS)
```

**Comment by froody** [comment]
Maybe related to https://github.com/pytorch/pytorch/issues/42132

**Comment by lw** [comment]
Still seeing it here: https://app.circleci.com/pipelines/github/pytorch/pytorch/196881/workflows/9ea1130f-a131-405b-a7e6-a5f8355d4370/jobs/6518861/steps

It didn't get skipped because that run occurred on a stack where these tests are being renamed (note to self: rename this issue once that stack lands). Thus it confirms that this issue persists.

**Comment by mrshenli** [comment]
Another instance: https://app.circleci.com/pipelines/github/pytorch/pytorch/197375/workflows/d5048196-a36b-4788-a1d3-6c138f688cf4/jobs/6545781/steps

**Comment by lw** [comment]
Thanks for the reminder! The stack has landed, so I updated the name of this issue to follow.

**Comment by pritamdamania87** [comment]
I'm wondering if this might have been resolved by https://github.com/pytorch/pytorch/pull/53170, will try to see if this still reproduces in CircleCI.

**Comment by pritamdamania87** [comment]
I ran this test 100 times in the circleci environment and couldn't reproduce it. I believe https://github.com/pytorch/pytorch/pull/53170 might have resolved this issue. @lw @mrshenli Feel free to re-open if the issue persists.
